training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala

```diff
@@ -73,13 +81,30 @@ object AdditiveModelTrainer {
 
     val paramsBC = sc.broadcast(params)
     var model = modelInitialization(sc, transformed, params)
-    for (i <- 1 to iterations) {
+    var loss = Double.MaxValue
+    for (i <- 1 to iterations
+         if loss >= params.loss.minLoss) {
       log.info(s"Iteration $i")
+      val sgdParams = SgdParams(
+        paramsBC,
+        sc.accumulator(0),
+        sc.accumulator(0))
       val modelBC = sc.broadcast(model)
-      model = sgdTrain(transformed, paramsBC, modelBC)
+      model = sgdTrain(transformed, sgdParams, modelBC)
       modelBC.unpersist()
-
-      TrainingUtils.saveModel(model, output)
+      val newLoss = sgdParams.loss.value/sgdParams.exampleCount.value
+      if (params.loss.useBestLoss) {
+         if (newLoss < loss) {
+           TrainingUtils.saveModel(model, output)
+           log.info(s"iterations $i useBestLoss ThisRoundLoss = $newLoss count = $sgdParams.exampleCount.value")
+           loss = newLoss
+         } else {
+           log.info(s"iterations $i loss $newLoss < best lost $loss count = $sgdParams.exampleCount.value")
+         }
+      } else {
+        TrainingUtils.saveModel(model, output)
+        log.info(s"iterations $i Loss = $newLoss count = $sgdParams.exampleCount.value")
+      }
     }
     model
```

Comment:
you are still returning the model in the last iteration, then the saveModel in `trainAndSaveToFileEarlySample` would overwrite whatever model you have saved with the last model